### PR TITLE
[Search] Be less restrictive with unknown keys on query endpoint

### PR DESCRIPTION
**Pull request by pedronastasi**
---
'@backstage/plugin-search-backend': minor
---

[Search] Be less restrictive with unknown keys on query endpoint by adding .passthrough() method to parse the requestSchema

```diff
-      const parseResult = requestSchema.safeParse(req.query);
+      const parseResult = requestSchema.passthrough().safeParse(req.query);
```
#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/mean-spiders-design.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+'@backstage/plugin-search-backend': major
```
Comment (emmaindal) [inline .changeset/mean-spiders-design.md:2]: I believe we should be able to keep this `minor` as its not a breaking change?

**Review by emmaindal — COMMENTED** [review]
Looks good to me, based on your discussion in this issue https://github.com/backstage/backstage/issues/13597, it would be nice if you can add a test for this use case too, maybe as part of https://github.com/backstage/backstage/blob/master/plugins/search-backend/src/service/router.test.ts.

@iamEAP can you have a final look too thanks!

File: .changeset/mean-spiders-design.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+'@backstage/plugin-search-backend': minor
+---
+
+Be less restrictive with unknown keys on query endpoint by adding .passthrough() method to parse the `requestSchema`
+
+```diff
+-      const parseResult = requestSchema.safeParse(req.query);
++      const parseResult = requestSchema.passthrough().safeParse(req.query);
+```
```
Comment (iamEAP) [inline .changeset/mean-spiders-design.md:10]: Hey @pedronastasi!  For context, this changeset entry eventually gets rolled up into [this changelog](https://github.com/backstage/backstage/blob/master/plugins/search-backend/CHANGELOG.md), which is often read by adopters as they upgrade packages.

With that audience in mind: I would recommend removing this diff line; your average adopter might not know/understand what it means, nor would they necessarily need that info.

File: .changeset/mean-spiders-design.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+---
+'@backstage/plugin-search-backend': minor
+---
+
+Be less restrictive with unknown keys on query endpoint by adding .passthrough() method to parse the `requestSchema`
```
Comment (iamEAP) [inline .changeset/mean-spiders-design.md:5]: Along the same lines (consider the audience), you could probably remove `by adding .passthrough() method to parse the requestSchema` here too; it's probably not useful information in the changelog.

**Review by iamEAP — COMMENTED** [review]
Thanks again @pedronastasi!  Some comments on the changelog.

I also want to re-up Emma's request for a test; we don't want to accidentally break this again! 
 Would you mind adding a test that checks that an unknown key sent in on the query endpoint is made available to the search engine's query method?  One quick case in `router.test.ts` ought to do.

File: plugins/search-backend/src/service/router.test.ts [hunk 4]
```diff
@@ -208,5 +211,55 @@ describe('createRouter', () => {
...
+          logger,
+        });
+        app = express().use(router);
+      });
+      it('pass through unknown keys', async () => {
+        const queryString =
+          'term=test&%5BdocType%5D%5B0%5D=Service&filters%5BdocType%5D%5B0%5D=filter1&unknownKey1%5B2%5D=unknownValue1&unknownKey1%5B3%5D=unknownValue2&unknownKey2=unknownValue1&pageCursor';
+        const response = await request(app).get(`/query?${queryString}`);
+        expect(response.status).toEqual(200);
+        expect(response.body).toMatchObject({ results: [expectedResult] });
```
Comment (iamEAP) [inline plugins/search-backend/src/service/router.test.ts:261]: Skimming through this test, it's a little unclear how checking the response body validates that unknown keys can be passed through.  It looks like most of that logic lives in the test setup.

To make it easier for future developers who stumble upon this code, maybe intsead of asserting on the result body, you could directly assert on the argument passed to the query?

Something like..

```ts
expect(mockSearchEngine.query).toHaveBeenCalledWith(expect.objectContaining({
  unknownKey1: [/* the expected value */],
  unknownKey2: '/* the expected value */'
});
```

File: plugins/search-backend/src/service/router.test.ts [hunk 5]
```diff
@@ -208,5 +211,55 @@ describe('createRouter', () => {
...
+    describe('Be less restrictive with unknown keys on query endpoint', () => {
+      const expectedResult = {
+        type: 'software-catalog',
+        document: {
+          text: 'foo',
+          title: 'bar baz',
+          location: '/catalog/default/component/example',
+        },
+      };
+      beforeAll(async () => {
```
Comment (iamEAP) [inline plugins/search-backend/src/service/router.test.ts:224]: If you're able to switch to just asserting on the query arguments, you could probably remove a lot of this test setup code too.

You could even consider forgoing this whole `describe()` block, and just adding the `pass through unknown keys` case into one of the existing describe blocks.  Perhaps `GET /query` block?

**Review by iamEAP — COMMENTED** [review]
Thanks again @pedronastasi!  Test looks correct; just a suggestion on clarity, then we can get this shipped.

**Comment by pedronastasi** [comment]
I made something wrong and somehow push Everyone's commit and because I couldn't go back to my previous state. I made a new brand PR. https://github.com/backstage/backstage/pull/13890

**Pull request closed without merging**